test/test_tap_insight.rb

```diff
@@ -0,0 +1,50 @@
+# encoding: UTF-8
+
+# TODO: I don't think that `test-unit-runner-tap` should be showing up
+# in backtraces, it should be filtered out.
+
+require 'stringio'
+require 'test/unit/ui/tap/ext/insight_testrunner'
+require 'json'
+require 'test/unit'
+
+class TestTapInsight < Test::Unit::TestCase
+
+  def test_failed
+    @test_case = Class.new(Test::Unit::TestCase) do
+      def test_success
+        assert_equal(3, 1 + 2)
+      end
+
+      def test_omit
+        omit
+        assert_equal(3, 1 + 2)
+      end
+
+      def test_pend
+        pend
+        assert_equal(3, 1 + 2)
+      end
+
+      def test_fail
+        assert_equal(3, 1 - 2)
+      end
+    end
+
+    @runner = Test::Unit::UI::Tap::InsightTestRunner.new(@test_case.suite, output: StringIO.new)
+    @result = @runner.start
+
+    assert_false(@result.passed?)
+    insight_json_report_path = File.join(File.dirname(__FILE__), '../test_unit_report.json')
+    assert_true(File.exist?(insight_json_report_path))
+
+    output = JSON.parse(File.read(insight_json_report_path))
+    test_case_results = output.select{|ele| ele["type"] == "test" }
+    assert_equal(test_case_results.count, 4)
+    assert_equal(test_case_results.find{|x| x["label"] == "test_success"}["status"], "pass")
+    assert_equal(test_case_results.find{|x| x["label"] == "test_fail"}["status"], "fail")
+    assert_equal(test_case_results.find{|x| x["label"] == "test_omit"}["status"], "skip")
+    assert_equal(test_case_results.find{|x| x["label"] == "test_pend"}["status"], "todo")
```

Comment:
Is it possible to verify that something was written to the `StringIO` instance? Does that make sense?